### PR TITLE
Expand dynamic drop tests for cases in #47949

### PR DESCRIPTION
**Pull request by matthewjasper**
Adds test cases for #47949

### CODE REVIEW

File: src/test/run-pass/drop/dynamic-drop.rs [hunk 1]
```diff
@@ -53,6 +53,21 @@ impl Allocator {
         data.push(true);
         Ptr(addr, self)
     }
+    // Creates a `Ptr<'_>` and checks that the allocated value is leaked if the
+    // the `failing_op` is in the list of exception.
+    //
+    // FIXME (#47949) We shouldn't need this, but we are sometimes leaking
+    // values when we panic in a destructor after evaluating an expression with
+    // `rustc_mir::builc::Builder::into`.
```
Comment (jonas-schievink) [inline src/test/run-pass/drop/dynamic-drop.rs]:
```suggestion
    // `rustc_mir::build::Builder::into`.
```

File: src/test/run-pass/drop/dynamic-drop.rs [hunk 2]
```diff
@@ -53,6 +53,21 @@ impl Allocator {
         data.push(true);
         Ptr(addr, self)
     }
+    // Creates a `Ptr<'_>` and checks that the allocated value is leaked if the
+    // the `failing_op` is in the list of exception.
```
Comment (jonas-schievink) [inline src/test/run-pass/drop/dynamic-drop.rs]:
```suggestion
    // `failing_op` is in the list of exception.
```

File: src/test/run-pass/drop/dynamic-drop.rs [hunk 3]
```diff
@@ -255,6 +270,66 @@ fn subslice_pattern_reassign(a: &Allocator) {
...
+
+fn panic_after_init_by_loop(a: &Allocator) {
+    // Panic in the drop of `r` can leak
+    let exceptions = vec![8];
+    a.alloc();
+    let p = a.alloc();
+    let q = loop {
+        a.alloc();
+        let r = a.alloc();
+        break a.alloc_leaked(exceptions);
```
Comment (nikomatsakis) [inline src/test/run-pass/drop/dynamic-drop.rs]: I'd prefer to have a FIXME here too -- something like

---

```
// FIXME(#XXX) we should not be leaking this, but we are
```

File: src/test/run-pass/drop/dynamic-drop.rs [hunk 4]
```diff
@@ -53,6 +53,21 @@ impl Allocator {
         data.push(true);
         Ptr(addr, self)
     }
+    // Creates a `Ptr<'_>` and checks that the allocated value is leaked if the
+    // `failing_op` is in the list of exception.
+    //
+    // FIXME (#47949) We shouldn't need this, but we are sometimes leaking
```
Comment (nikomatsakis) [inline src/test/run-pass/drop/dynamic-drop.rs]: For some reason, I found this kind of confusing. Maybe something like this:

```
FIXME(#47949) -- Each use of this indicates a bug: basically, a value that winds up getting leaked but shouldn't be.
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good, r=me, but can we tweak the comments?

**Comment by nikomatsakis** [comment]
r=me -- @matthewjasper did you want to "FIXUP" that commit, thogh?

**Pull request merged**